### PR TITLE
Refactory gang management code.

### PR DESCRIPTION
**Pull request by yydzero**
This commit is a first step to abstract gang management (and dispatch later) code in
order to allow different implementation easier.

@hlinnaka As you are also changing this part of code, would you please help to review this?

### CODE REVIEW

**Comment by hlinnaka** [comment]
What's the point of abstracting the gang management this way?

**Comment by yydzero** [comment]
The purpose is to allow different implementations of dispatch/gang code easier without change calling code too much (sth. like MemoryContext or polymorphism in OOP), and also easier to switch between different implementations for different purpose (like OLTP vs. OLAP vs. mixed workload etc).  

Historically there is dispatch agent code, while it is spread to many places and hard to maintain.

**Comment by atris** [comment]
Just checking, the main aim of the PR is to abstract out the gang and dispatch code out into a separate manager with an API which is consumed?

Does this lead to definition of an abstract API that can be used to treat the underlying gang manager as a black box? Something like a gang and dispatch manager interface?

**Comment by yydzero** [comment]
The main goal is to provide an interface for dispatch/gang and caller will
use it as a black box.  This first commit is far from that, while a good
step to start discussion and sourcing various thoughts before going too
further.

On Tue, Feb 16, 2016 at 2:42 AM, Atri Sharma <notifications@github.com>
wrote:

> Just checking, the main aim of the PR is to abstract out the gang and
> dispatch code out into a separate manager with an API which is consumed?
>
> Does this lead to definition of an abstract API that can be used to treat
> the underlying gang manager as a black box? Something like a gang and
> dispatch manager interface?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/373#issuecomment-184342896>.
>



-- 
Best Regards,
Yandong

**Comment by yydzero** [comment]
Any more comments?

File: src/backend/cdb/cdbexplain.c [hunk 1]
```diff
@@ -24,6 +24,7 @@
 #include "utils/memutils.h"             /* MemoryContextGetPeakSpace() */
 #include "cdb/memquota.h"
 #include "cdb/cdbgang.h"
+#include "cdb/cdbslicetable.h"
```
Comment (gcaragea) [inline src/backend/cdb/cdbexplain.c]: Probably not needed.

File: gpAux/extensions/gpfdist/configure [hunk 2]
```diff
@@ -3304,7 +3304,7 @@ if test -n "$APR_CONFIG"; then
   LIBS="$LIBS $apr_link_ld_libs"
 fi
 
-CFLAGS="-Werror $CFLAGS"
+CFLAGS="-Werror -Wno-deprecated -Wno-parentheses-equality $CFLAGS"
```
Comment (gcaragea) [inline gpAux/extensions/gpfdist/configure]: These are unrelated changes AFAIK, please remove from PR.

File: src/backend/cdb/cdbgang.c [hunk 3]
```diff
@@ -2901,90 +2619,138 @@ isTargetPortal(const char *p1, const char *p2)
...
 
-#ifdef USE_ASSERT_CHECKING
-/**
- * Assert that slicetable is valid. Must be called after ExecInitMotion, which sets up the slice table
- */
-void
-AssertSliceTableIsValid(SliceTable *st, struct PlannedStmt *pstmt)
+
+// ======================================================
+// TODO: Move below methods to proper places.
```
Comment (gcaragea) [inline src/backend/cdb/cdbgang.c]: Are you planning to do this before pushing?

**Comment by gcaragea** [comment]
Can you give us a preview of the plans you have for a different gang manager mechanism? That would help give context to these changes. Thanks.

File: gpAux/extensions/gpfdist/configure (hunk 2 above)
Comment (yydzero) [inline gpAux/extensions/gpfdist/configure]: Will do.

File: src/backend/cdb/cdbexplain.c (hunk 1 above)
Comment (yydzero) [inline src/backend/cdb/cdbexplain.c]: Will remove it

File: src/backend/cdb/cdbgang.c (hunk 3 above)
Comment (yydzero) [inline src/backend/cdb/cdbgang.c]: Will do it in late commit.

**Comment by yydzero** [comment]
Some rough idea of other mechanisms, like dispatcher agent, flexible gang resource sharing across sessions, more gang types etc.

By refactoring those code (dispatcher/gang etc), we will open the door for trying with more mechanisms for mixed work load without impact existing mechanism.  This commit is very early first step towards that. 

Also would like to hear any feedback about simple & flexible method to allow different implementations in C for same interface.

**Comment by hlinnaka** [comment]
On 18/02/16 05:27, Yandong Yao wrote:
> Some rough idea of other mechanisms, like dispatcher agent, flexible
> gang resource sharing across sessions, more gang types etc.
>
> By refactoring those code (dispatcher/gang etc), we will open the
> door for trying with more mechanisms for mixed work load without
> impact existing mechanism.  This commit is very early first step
> towards that.

I think we should drop this for now, and do this as part of whatever 
alternative mechanism you have in mind. It might turn out that for those 
things, you need a different code structure and end up rewriting much of 
this anyway. As long as we have only one dispatcher implementation, I 
don't think this is an improvement in terms of readability.

> Also would like to hear any feedback about simple & flexible method
> to allow different implementations in C for same interface.

Well, a struct with function pointers, like you did in this patch, is 
probably the most straightforward way. Or if it's enough to choose one 
implementation at startup time, you could use global variables instead 
of a struct. But the question is whether we need any of that at all.

- Heikki

**Comment by yydzero** [comment]
On Thu, Feb 18, 2016 at 7:25 PM, Heikki Linnakangas <
notifications@github.com> wrote:

> On 18/02/16 05:27, Yandong Yao wrote:
> > Some rough idea of other mechanisms, like dispatcher agent, flexible
> > gang resource sharing across sessions, more gang types etc.
> >
> > By refactoring those code (dispatcher/gang etc), we will open the
> > door for trying with more mechanisms for mixed work load without
> > impact existing mechanism. This commit is very early first step
> > towards that.
>
> I think we should drop this for now, and do this as part of whatever
> alternative mechanism you have in mind. It might turn out that for those
> things, you need a different code structure and end up rewriting much of
> this anyway. As long as we have only one dispatcher implementation, I
> don't think this is an improvement in terms of readability.
>

Though I don't agree with you, as there are different level of
abstraction/layers (as a bunch of function calls), we definitely could make
some layer has no relationship with top layer and make things much clearer,
like the sole dispatching code which send serialized string/request to a
bunch of QEs and get result with error/cancel handling has no need to
understand the content of the request, whether it is a DTX command, planned
statement or string query.  While I don't want to spend lots of more time
to discuss whether it is needed to do this refactory, I will work on it on
my local branch and find a proper time to do full PR which might not happen
in near time compared with my velocity/time on story.


>
> > Also would like to hear any feedback about simple & flexible method
> > to allow different implementations in C for same interface.
>
> Well, a struct with function pointers, like you did in this patch, is
> probably the most straightforward way. Or if it's enough to choose one
> implementation at startup time, you could use global variables instead
> of a struct. But the question is whether we need any of that at all.
>
> - Heikki
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/373#issuecomment-185673015>.
>



-- 
Best Regards,
Yandong

**Comment by hlinnaka** [comment]
On 19/02/16 05:11, Yandong Yao wrote:
> On Thu, Feb 18, 2016 at 7:25 PM, Heikki Linnakangas <
> notifications@github.com> wrote:
>
>> I think we should drop this for now, and do this as part of whatever
>> alternative mechanism you have in mind. It might turn out that for those
>> things, you need a different code structure and end up rewriting much of
>> this anyway. As long as we have only one dispatcher implementation, I
>> don't think this is an improvement in terms of readability.
>
> Though I don't agree with you, as there are different level of
> abstraction/layers (as a bunch of function calls), we definitely could make
> some layer has no relationship with top layer and make things much clearer,
> like the sole dispatching code which send serialized string/request to a
> bunch of QEs and get result with error/cancel handling has no need to
> understand the content of the request, whether it is a DTX command, planned
> statement or string query.  While I don't want to spend lots of more time
> to discuss whether it is needed to do this refactory, I will work on it on
> my local branch and find a proper time to do full PR which might not happen
> in near time compared with my velocity/time on story.

Yeah, I could certainly get on board refactoring the dispatch interface 
to be more clear in general. I just didn't think this particular 
abstraction was useful. Not sure what would be, I haven't given it any 
thought.

Thanks!

- Heikki
